web/src/message_events.js

```diff
@@ -35,6 +37,129 @@ import * as unread_ops from "./unread_ops";
 import * as unread_ui from "./unread_ui";
 import * as util from "./util";
 
+export function update_views_filtered_on_message_property(message_ids, term_type) {
+    // NOTE: Call this function after updating the message property locally.
+    assert(!term_type.includes("not-"));
+
+    // List of narrow terms whose msg list doesn't get updated elsewhere but
+    // can be applied locally.
+    const supported_term_types = [
+        "has-image",
+        "has-link",
+        "has-reaction",
+        "has-attachment",
+        "is-starred",
+        "is-unread",
+        "is-mentioned",
+        "is-alerted",
+        // TODO: Implement support for these terms.
+        // "is-followed",
+    ];
+
+    if (message_ids.length === 0 || !supported_term_types.includes(term_type)) {
+        return;
+    }
+
+    for (const msg_list of message_lists.all_rendered_message_lists()) {
+        const filter = msg_list.data.filter;
+        const filter_term_types = filter.sorted_term_types();
+        if (
+            // Check if current filter relies on the changed message property.
+            !filter_term_types.includes(term_type) &&
+            !filter_term_types.includes(`not-${term_type}`)
+        ) {
+            continue;
+        }
+
+        if (!filter.can_apply_locally()) {
+            // We can't apply the update locally, so we need to rerender the view.
+            const current_selected_id = msg_list.selected_id();
+            message_view.show(filter.terms(), {
+                then_select_id: current_selected_id,
+                trigger: "message property update",
+                force_rerender: true,
+            });
+            continue;
+        }
+
+        // We need the message objects to determine if they match the filter.
+        const messages_to_fetch = [];
+        const messages = [];
+        for (const message_id of message_ids) {
+            const message = message_store.get(message_id);
+            if (message !== undefined) {
+                messages.push(message);
+            } else {
+                const first_id = msg_list.first().id;
```

Comment:
I feel like somewhere around this point, we could probably have a check for the case for cases of the form "The view includes `is:starred`, the change is that a message is no longer starred, and the message isn't already in the view, so this change can't possibly add the message to the view".

Reply:
Added the relevant logic.